### ISSUE TITLE
Error on local client startup if offline (MacOS)

### ISSUE THREAD

**Issue by velicanu** [dask/distributed#3916]
**What happened**:
When connected to the internet code runs. When not connected to the internet:
```
python test.py
Traceback (most recent call last):
  File "/Users/dvelicanu/git/expenses/venv/lib/python3.7/site-packages/toolz/functoolz.py", line 456, in memof
    return cache[k]
KeyError: ('8.8.8.8', 80)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/dvelicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/utils.py", line 135, in _get_ip
    sock.connect((host, port))
OSError: [Errno 51] Network is unreachable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test.py", line 1, in <module>
    from distributed import Client
  File "/Users/dvelicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/__init__.py", line 3, in <module>
    from .actor import Actor, ActorFuture
  File "/Users/dvelicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/actor.py", line 6, in <module>
    from .client import Future, default_client
  File "/Users/dvelicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/client.py", line 43, in <module>
    from .batched import BatchedSend
  File "/Users/dvelicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/batched.py", line 8, in <module>
    from .core import CommClosedError
  File "/Users/dvelicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/core.py", line 19, in <module>
    from .comm import (
  File "/Users/dvelicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/comm/__init__.py", line 26, in <module>
    _register_transports()
  File "/Users/dvelicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/comm/__init__.py", line 17, in _register_transports
    from . import inproc
  File "/Users/dvelicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/comm/inproc.py", line 75, in <module>
    global_manager = Manager()
  File "/Users/dvelicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/comm/inproc.py", line 39, in __init__
    self.ip = get_ip()
  File "/Users/dvelicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/utils.py", line 159, in get_ip
    return _get_ip(host, port, family=socket.AF_INET)
  File "/Users/dvelicanu/git/expenses/venv/lib/python3.7/site-packages/toolz/functoolz.py", line 460, in memof
    cache[k] = result = func(*args, **kwargs)
  File "/Users/dvelicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/utils.py", line 145, in _get_ip
    socket.gethostname(), port, family, socket.SOCK_DGRAM, socket.IPPROTO_UDP
  File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/socket.py", line 748, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 8] nodename nor servname provided, or not known
```

**What you expected to happen**:
Expecting the local client to work whether or not the internet is connected.
```
python test.py
<Client: 'tcp://127.0.0.1:51784' processes=4 threads=12, memory=17.18 GB>
```

**Minimal Complete Verifiable Example**:

```python
from distributed import Client

if __name__ == "__main__":
    client = Client()
    print(client)
```

**Anything else we need to know?**:
I could not reproduce this bug on a Windows machine. But I have been able to reproduce it on two different Mac machines.

**Environment**:

- Dask version: dask==2.19.0 , distributed==2.19.0
- Python version: Python 3.7.4
- Operating System: macOS Mojave 10.14.6 (18G95)
- Install method (conda, pip, source): pip

**Comment by quasiben** [comment]
Hmm, that is rather strange.  Perhaps there is something odd in the /etc/hosts file or maybe DNS lookup is mucking around with things -- this can happen if you are on a VPN and using 8.8.8.8.  For now could you try the following

```python
In [1]: from distributed import Client, LocalCluster

In [2]: cluster = LocalCluster(ip="127.0.0.1")

In [2]: client = Client(cluster)
```

**Comment by velicanu** [comment]
Just tried that out. It turns out it crashes on the import statement, I didn't even get to `[2]`!

This is with WIFI turned OFF. With internet connection there's no problem.

```
$ python
Python 3.7.5 (default, Nov  1 2019, 02:16:32)
[Clang 11.0.0 (clang-1100.0.33.8)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from distributed import Client, LocalCluster
Traceback (most recent call last):
  File "/Users/velicanu/git/expenses/venv/lib/python3.7/site-packages/toolz/functoolz.py", line 456, in memof
    return cache[k]
KeyError: ('8.8.8.8', 80)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/velicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/utils.py", line 136, in _get_ip
    sock.connect((host, port))
OSError: [Errno 51] Network is unreachable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/velicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/__init__.py", line 3, in <module>
    from .actor import Actor, ActorFuture
  File "/Users/velicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/actor.py", line 6, in <module>
    from .client import Future, default_client
  File "/Users/velicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/client.py", line 43, in <module>
    from .batched import BatchedSend
  File "/Users/velicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/batched.py", line 8, in <module>
    from .core import CommClosedError
  File "/Users/velicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/core.py", line 19, in <module>
    from .comm import (
  File "/Users/velicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/comm/__init__.py", line 26, in <module>
    _register_transports()
  File "/Users/velicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/comm/__init__.py", line 17, in _register_transports
    from . import inproc
  File "/Users/velicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/comm/inproc.py", line 75, in <module>
    global_manager = Manager()
  File "/Users/velicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/comm/inproc.py", line 39, in __init__
    self.ip = get_ip()
  File "/Users/velicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/utils.py", line 160, in get_ip
    return _get_ip(host, port, family=socket.AF_INET)
  File "/Users/velicanu/git/expenses/venv/lib/python3.7/site-packages/toolz/functoolz.py", line 460, in memof
    cache[k] = result = func(*args, **kwargs)
  File "/Users/velicanu/git/expenses/venv/lib/python3.7/site-packages/distributed/utils.py", line 146, in _get_ip
    socket.gethostname(), port, family, socket.SOCK_DGRAM, socket.IPPROTO_UDP
  File "/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/socket.py", line 748, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 8] nodename nor servname provided, or not known
>>>
```

**Comment by quasiben** [comment]
Can you try the following:

``python
import socket
s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
s.connect(("8.8.8.8", 80))
print(s.getsockname()[0])
s.close()
```

Can you also post the contents of `/etc/hosts` and `/etc/resolv.conf`

**Comment by velicanu** [comment]
Without internet:
```
$ python
Python 3.7.5 (default, Nov  1 2019, 02:16:32)
[Clang 11.0.0 (clang-1100.0.33.8)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import socket
>>> s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
>>> s.connect(("8.8.8.8", 80))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 51] Network is unreachable
>>> print(s.getsockname()[0])
0.0.0.0
>>> s.close()
>>>
```

With internet:
```
$ python
Python 3.7.5 (default, Nov  1 2019, 02:16:32)
[Clang 11.0.0 (clang-1100.0.33.8)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import socket
>>> s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
>>> s.connect(("8.8.8.8", 80))
>>> print(s.getsockname()[0])
192.168.0.119
>>> s.close()
>>>
```

Contents of `/etc/hosts`:
```bash
$ cat /etc/hosts
##
# Host Database
#
# localhost is used to configure the loopback interface
# when the system is booting.  Do not change this entry.
##
127.0.0.1	localhost
255.255.255.255	broadcasthost
::1             localhost
# Added by Docker Desktop
# To allow the same kube context to work on the host and the container:
127.0.0.1 kubernetes.docker.internal
# End of section
```


`/etc/resolv.conf` only exists when the internet is connected:
```bash
$ cat /etc/resolv.conf
#
# macOS Notice
#
# This file is not consulted for DNS hostname resolution, address
# resolution, or the DNS query routing mechanism used by most
# processes on this system.
#
# To view the DNS configuration used by this system, use:
#   scutil --dns
#
# SEE ALSO
#   dns-sd(1), scutil(8)
#
# This file is automatically generated.
#
nameserver 192.168.0.1
nameserver 0.0.0.0
```

Without internet:
```bash
$ cat /etc/resolv.conf
cat: /etc/resolv.conf: No such file or directory
```

**Comment by jonashaag** [comment]
Have the same problem. Please tell if I can be of any help debugging this.

**Comment by jsignell** [comment]
Closed by #3991

**Comment by crusaderky** [comment]
PR does not fix the issue for me. 

Stack trace after the PR:

```
Traceback (most recent call last):
  File "/Users/crusaderky/miniconda3/envs/commrisk/lib/python3.8/site-packages/distributed/utils.py", line 136, in _get_ip
    sock.connect((host, port))
OSError: [Errno 51] Network is unreachable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/crusaderky/miniconda3/envs/commrisk/lib/python3.8/site-packages/distributed/__init__.py", line 3, in <module>
    from .actor import Actor, ActorFuture
  File "/Users/crusaderky/miniconda3/envs/commrisk/lib/python3.8/site-packages/distributed/actor.py", line 6, in <module>
    from .client import Future, default_client
  File "/Users/crusaderky/miniconda3/envs/commrisk/lib/python3.8/site-packages/distributed/client.py", line 43, in <module>
    from .batched import BatchedSend
  File "/Users/crusaderky/miniconda3/envs/commrisk/lib/python3.8/site-packages/distributed/batched.py", line 8, in <module>
    from .core import CommClosedError
  File "/Users/crusaderky/miniconda3/envs/commrisk/lib/python3.8/site-packages/distributed/core.py", line 19, in <module>
    from .comm import (
  File "/Users/crusaderky/miniconda3/envs/commrisk/lib/python3.8/site-packages/distributed/comm/__init__.py", line 26, in <module>
    _register_transports()
  File "/Users/crusaderky/miniconda3/envs/commrisk/lib/python3.8/site-packages/distributed/comm/__init__.py", line 17, in _register_transports
    from . import inproc
  File "/Users/crusaderky/miniconda3/envs/commrisk/lib/python3.8/site-packages/distributed/comm/inproc.py", line 75, in <module>
    global_manager = Manager()
  File "/Users/crusaderky/miniconda3/envs/commrisk/lib/python3.8/site-packages/distributed/comm/inproc.py", line 39, in __init__
    self.ip = get_ip()
  File "/Users/crusaderky/miniconda3/envs/commrisk/lib/python3.8/site-packages/distributed/utils.py", line 160, in get_ip
    return _get_ip(host, port, family=socket.AF_INET)
  File "cytoolz/functoolz.pyx", line 476, in cytoolz.functoolz._memoize.__call__
  File "/Users/crusaderky/miniconda3/envs/commrisk/lib/python3.8/site-packages/distributed/utils.py", line 145, in _get_ip
    addr_info = socket.getaddrinfo(
  File "/Users/crusaderky/miniconda3/envs/commrisk/lib/python3.8/socket.py", line 918, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 8] nodename nor servname provided, or not known
```

**Comment by jsignell** [comment]
Hmm maybe we are not catching all the right kinds of errors in that PR.

**Comment by crusaderky** [comment]
It's weird. I can still replicate the issue with the old code, but I can't after pip-installing the distributed git tip.
It looks like in my retest yesterday I botched the upgrade for some reason...

**Comment by velicanu** [comment]
This issue is fixed for me too on version `distributed==2.21.0+13.g7cdba1d7` (current master) even though I still see it when I pip install distributed and on version `distributed==2.21.0`

So code-wise it seems fixed.

**Comment by jsignell** [comment]
Ok. I am closing this again since it seems likely that it is fixed.
